rust/src/librarydb.rs

```diff
@@ -0,0 +1,504 @@
+//!  * it is thread safe
+//!  * shadowing respects library order and uses the full file path,
+//!    the original shadowed libraries based on the longest base path
+//!
+//!
+//! # FFI
+//!
+//! [`stracciatella::c::librarydb`] contains a C interface for this module.
+//!
+//! [`stracciatella::c::librarydb`]: ../c/librarydb/index.html
+
+use std::cmp;
+use std::collections::HashSet;
+use std::fmt;
+use std::fs::File;
+use std::io::{self, Seek, SeekFrom};
+use std::ops;
+use std::path::{Component, Path, PathBuf};
+use std::sync::{Arc, Mutex, MutexGuard, RwLock};
+
+use crate::file_formats::slf::{SlfEntryState, SlfHeader};
+
+/// Thread safe library database.
+#[derive(Debug)]
+pub struct LibraryDB {
+    inner: Mutex<LibraryDBInner>,
+}
+
+/// Library database.
+#[derive(Debug)]
+pub struct LibraryDBInner {
+    /// Thread safe libraries.
+    arc_libraries: Vec<Arc<RwLock<Library>>>,
+}
+
+/// Library.
+/// Keeps the library file open for reading and a list of entries that are ok.
+/// The library file contains the data of each entry.
+#[derive(Debug)]
+pub struct Library {
+    /// Real path of the library file for display purposes.
+    library_path: PathBuf,
+    /// Library file open for reading.
+    library_file: File,
+    /// Base path of the entries in the library.
+    base_path: String,
+    /// List of ok entries in the library.
+    entries: Vec<LibraryEntry>,
+}
+
+/// Library entry.
+/// Represents a file in a library.
+#[derive(Debug)]
+pub struct LibraryEntry {
+    /// File path without the base path of the library.
+    file_path: String,
+    /// Start of the file data in the library file.
+    data_start: u64,
+    /// End of the file data in the library file.
+    data_end: u64,
+}
+
+/// Library file.
+/// Provides access to the data of a library entry through Read and Seek.
+#[derive(Debug)]
+pub struct LibraryFile {
+    /// The index of the entry of this file.
+    index: usize,
+    /// The thread safe library of this file.
+    arc_library: Arc<RwLock<Library>>,
+    /// Current position in the library file.
+    position: u64,
+}
+
+impl LibraryDB {
+    /// Constructor.
+    pub fn new() -> Self {
+        Self {
+            inner: Mutex::new(LibraryDBInner::new()),
+        }
+    }
+
+    /// Returns thread safe guard that grants direct access.
+    /// While the guard lives, no one else has access to the inner library database.
+    pub fn locked(&self) -> MutexGuard<LibraryDBInner> {
+        self.inner.lock().unwrap()
+    }
+
+    /// Opens and adds a library at the end of library database.
+    pub fn add_library(&mut self, data_dir: &Path, library: &Path) -> io::Result<()> {
+        self.locked().add_library(data_dir, library)
+    }
+
+    /// Opens a library file for reading.
+    /// The file must be dropped before the library database is dropped.
+    pub fn open_file(&self, path: &str) -> io::Result<LibraryFile> {
+        self.locked().open_file(path)
+    }
+}
+
+impl LibraryDBInner {
```

Comment:
Do we need this `Inner` construct?  As long as we only have `arc_libraries` we should be good without it right?

Reply:
It splits simple access (`LibraryDB`) and guarded access (provides access to `LibraryDBInner`).
I'm not happy with the name but it is needed because we have custom functions (I can't do `impl Vec<Arc<RwLock<Library>>>`).